Docs for Solving Wellbore Problems with Poroplastic Solver (#1475)
GEOSX/GEOSX: src/docs/sphinx/tutorials/step12/Tutorial.rst

.. _TutorialPoroPlasticity:

####################################################
Tutorial 12: PoroMechanical Models for Wellbore Problems
####################################################

**Context**

> The main objective of this tutorial is to demonstrate how to use the internal wellbore mesh generator and poromechanical solvers in GEOSX to tackle wellbore problems in porous media. In this tutorial, a poroplastic model is applied to find the solution of rock deformation within the vicinity of a vertical wellbore, considering elastoplastic deformation, fluid diffusion and poromechanical coupling effect. To do so, a single phase flow solver is fully coupled with a Lagrangian mechanics solver in GEOSX and the Extended Drucker-Prager model (see :ref:`TwoInvariantPlasticity`) is chosen as the material model for the solid domain. We first solve this problem with the existing poroelastic solver and verify the modeling results with the corresponding analytical solutions. Then, the verified case is employed to test the newly implemented poroplasticity solver in GEOSX, whose results are compared with the ones obtained from the poroelastic case to highlight the impact of plasticity in this specific problem.

Reviewer: ```suggestion
The main objective of this tutorial is to demonstrate how to use the internal wellbore mesh generator and poromechanical solvers in GEOSX to tackle wellbore problems in porous media. In this tutorial, a poroplastic model is applied to find the solution of rock deformation within the vicinity of a vertical wellbore, considering elastoplastic deformation, fluid diffusion and poromechanical coupling effect. To do so, a single phase flow solver is fully coupled with a Lagrangian mechanics solver and the Extended Drucker-Prager model (see :ref:`TwoInvariantPlasticity`) is chosen as the material model for the solid domain. We first solve this problem with a poroelastic model and verify the modeling results with the corresponding analytical solutions. Then, the verified case is modified to test a poroplastic version, whose results are compared with the ones obtained from the poroelastic case to highlight the impact of plasticity in this specific problem.
```

 

**Objectives**

At the end of this tutorial you will know:

  - how to construct meshes for wellbore problems with the internal wellbore mesh generator,
  - how to specify initial and boundary conditions, such as reservoir properties, in-situ stresses, mixed loading (mechanical and fluid) at wellbore wall and far-field constraints,
  - how to use multiple solvers in GEOSX for predicting poroplastic deformations in the near wellbore region.

**Input file**

This tutorial uses no external input files and everything required is
contained within a single GEOSX input file. 

The xml input file for the test case with poroelasticity is located at:

.. code-block:: console

  examples/plasticity/WellboreProblem_PoroElastic.xml

The xml input file for the test case with poroplasticity is located at:

.. code-block:: console

  examples/plasticity/WellboreProblem_PoroDruckerPrager.xml

------------------------------------------------------------------
Description of the case
------------------------------------------------------------------

> 

Reviewer: General comments:

- It is not necessary to ever say "in GEOSX."  I would be impressed if the user was unaware of the code they are working in :-)
- We do not have separate Poroelastic and Poroplastic *solvers*.  These are one solver with different model specifications.  I would refer to them always as the "poroelastic version" and "poroplastic version" of the problem.

> We simulate the wellbore problem subjected to anisotropic horizontal stress (:math:`\Sigma_h` and :math:`\Sigma_H`) and vertical stress (:math:`\Sigma_v`), as shown in Fig.1. This is a vertical wellbore, which is drilled in a porous medium. By changing the wellbore supporting pressure (could be traction and/or fluid loading), the mechanical deformation of the reservoir rock will be induced and evolve with time, due to fluid diffusion and coupling effect. Considering inelastic constitutive behavior, reservoir rock in the near wellbore region will experience elastoplastic deformation and a plastic zone will be developed and expand with time. To setup the base case, the poroelastic solver is employed to find the poroelastic solutions of this wellbore problem, which are verified with the analytical solution `(Detournay and Cheng, 1993)  <https://www.sciencedirect.com/science/article/pii/B9780080406152500113>`__ from the literature. Following that, a poroplastic solver in GEOSX is built and used to obtain the temporal and spatial solutions of pore pressure, displacement and stress fields around the wellbore, considering induced plastic deformation. 

Reviewer: ```suggestion
We simulate the wellbore problem subjected to anisotropic horizontal stress (:math:`\Sigma_h` and :math:`\Sigma_H`) and vertical stress (:math:`\Sigma_v`), as shown in Fig.1. This is a vertical wellbore, which is drilled in a porous medium. By changing the wellbore supporting pressure, the mechanical deformation of the reservoir rock will be induced and evolve with time, due to fluid diffusion and coupling effect. Considering inelastic constitutive behavior, the reservoir rock in the near wellbore region will experience elastoplastic deformation and a plastic zone will be developed and expand with time. To setup the base case, a poroelastic version is employed to find the poroelastic solutions of this wellbore problem, which are verified with the analytical solution `(Detournay and Cheng, 1993)  <https://www.sciencedirect.com/science/article/pii/B9780080406152500113>`__ from the literature. Following that, a poroplastic version is built and used to obtain the temporal and spatial solutions of pore pressure, displacement and stress fields around the wellbore, considering induced plastic deformation. 
```

to conform with the wellbore geometry. This mesh is defined as a cell block with the name
``cb1``.

.. literalinclude:: ../../../../../examples/plasticity/WellboreProblem_PoroDruckerPrager.xml
    :language: xml
    :start-after: <!-- SPHINX_WELLBORE_MESH -->
    :end-before: <!-- SPHINX_WELLBORE_MESH_END -->

Solver: setting up the solid mechanics solver
-----------------------------------------------------------

GEOSX is a multi-physics platform. Different combinations of
physics solvers available in the code can be applied
in different regions of the domain and be functional at different stages of the simulation.
The ``Solvers`` tag in the XML file is used to list and parameterize these solvers.

To specify a coupling between two different solvers, we define and characterize each single-physics solver separately.
Then, we customize a *coupling solver* between these single-physics
solvers as an additional solver.
This approach allows for generality and flexibility in constructing multi-physics solvers.
The order of specifying these solvers is not restricted in GEOSX.

> Note that end-users should give each single-physic solver a meaningful and distinct name, as GEOSX will recognize these single-physics solvers based on their customized names and create user-expected coupling.

Reviewer: ```suggestion
Note that end-users should give each single-physics solver a meaningful and distinct name, as GEOSX will recognize these single-physics solvers based on their customized names and create user-expected coupling.
```

> As demonstrated in this tutorial, to setup a poro-mechanical coupling, we need to effectively define three different solvers in the XML file:

Reviewer: ```suggestion
As demonstrated in this tutorial, to setup a poromechanical coupling, we need to define three different solvers in the XML file:
```

- the small strain Lagrangian mechanics solver, a solver of type ``SolidMechanics_LagrangianFEM`` called here ``mechanicsSolver`` (more information here: :ref:`SolidMechanicsLagrangianFEM`),

.. literalinclude:: ../../../../../examples/plasticity/WellboreProblem_PoroDruckerPrager.xml
  :language: xml
  :start-after: <!-- SPHINX_WELLBORE_MECHANICALSOLVER -->
  :end-before: <!-- SPHINX_WELLBORE_MECHANICALSOLVER_END -->

- the single-phase flow solver, a solver of type ``SinglePhaseFVM`` called here ``SinglePhaseFlowSolver`` (more information on these solvers at :ref:`SinglePhaseFlow`),

.. literalinclude:: ../../../../../examples/plasticity/WellboreProblem_PoroDruckerPrager.xml
  :language: xml
  :start-after: <!-- SPHINX_WELLBORE_SINGLEPHASEFVM -->
  :end-before: <!-- SPHINX_WELLBORE_SINGLEPHASEFVM_END -->

> - the coupling solver (``Poroelastic``) that will bind the two single-physics solvers above, which is named as ``PoroelasticitySolver`` (more information at :ref:`PoroelasticSolver`).

Reviewer: Poromechanics may be a better name given we are solving a plastic problem.

Author: Right, same solver (SinglePhasePoromechanics) is used for solving both poroelastic and poroplastic problems, whose major difference is the material model. To avoid confusion, PoromechanicsSolver should be used in this tutorial.

.. literalinclude:: ../../../../../examples/plasticity/WellboreProblem_PoroDruckerPrager.xml
  :language: xml
  :start-after: <!-- SPHINX_WELLBORE_POROELASTICSOLVER -->
  :end-before: <!-- SPHINX_WELLBORE_POROELASTICSOLVER_END -->

The two single-physics solvers are parameterized as explained
in their corresponding documents. 

In this tutorial, let us focus on the coupling solver.

> This solver (``PoroelasticitySolver``) uses a set of attributes that specifically describe the coupling process within a poromechanical framework.

Reviewer: PoromechanicsSolver

For instance, we must point this solver to the designated fluid solver (here: ``SinglePhaseFlowSolver``) and solid solver (here: ``mechanicsSolver``).

> These solvers are forced to interact through the ``porousMaterialNames="{porousRock}"`` with all the constitutive models. We specify the the discretization method (``FE1``, defined in the ``NumericalMethods`` section), and the target regions (here, we only have one, ``Omega``).

Reviewer: ```suggestion
These solvers are forced to interact through the ``porousMaterialNames="{porousRock}"`` with all the constitutive models. We specify the discretization method (``FE1``, defined in the ``NumericalMethods`` section), and the target regions (here, we only have one, ``Omega``).
```

> And more parameters are required to characterize a coupling procedure (more information at :ref:`PoroelasticSolver`). In this way, the two single-physics solvers will be simultaneously called and executed for solving the wellbore problem here.

Reviewer: English style point: sentences should never begin with `And`.  You can simply say, "More parameters ..." or "In addition, extra parameters ..."

Author: Agree

please see the dedicated :ref:`FiniteElement` section.

The finite volume method requires the specification of a discretization scheme.
Here, we use a two-point flux approximation scheme (``singlePhaseTPFA``), as described in the dedicated documentation (found here: :ref:`FiniteVolume`).

.. literalinclude:: ../../../../../examples/plasticity/WellboreProblem_PoroDruckerPrager.xml
  :language: xml
  :start-after: <!-- SPHINX_WELLBORE_NUMERICAL -->
  :end-before: <!-- SPHINX_WELLBORE_NUMERICAL_END -->

Constitutive model: defining material properties with constitutive laws
-----------------------------------------------------------

For this test problem, the solid and fluid materials are named as ``rock`` and ``water`` respectively, whose mechanical properties are specified in the ``Constitutive`` section. In this tutorial, different material models, linear elastic isotropic model (see :ref:`LinearElasticIsotropic`) and Extended Drucker-Prager model (see :ref:`TwoInvariantPlasticity`), are used to solve the mechanical deformation, which is the only difference between the poroelastic and poroplastic cases in this tutorial.

> For the poroelastic case, linear elastic isotropic solid model ``PoroElasticIsotropic`` is used to describe the linear elastic isotropic response of ``rock`` to loading. And the single-phase fluid model ``CompressibleSinglePhaseFluid`` is selected to simulate the transport phenomenon of ``water`` upon injection:

Reviewer: ```suggestion
For the poroelastic case, linear elastic isotropic solid model ``PoroElasticIsotropic`` is used to describe the linear elastic isotropic response of ``rock`` to loading. And the single-phase fluid model ``CompressibleSinglePhaseFluid`` is selected to simulate the flow of ``water`` upon injection:
```

.. literalinclude:: ../../../../../examples/plasticity/WellboreProblem_PoroElastic.xml
    :language: xml
    :start-after: <!-- SPHINX_WELLBORE_MATERIAL -->
    :end-before: <!-- SPHINX_WELLBORE_MATERIAL_END -->

> For the poroplastic case, Extended Drucker-Prager model ``PoroExtendedDruckerPrager`` is used to simulate the elastoplastic behavior of ``rock``. And the single-phase fluid model ``CompressibleSinglePhaseFluid`` is employed to handle the storage and transport of ``water``: 

Reviewer: ```suggestion
For the poroplastic case, Extended Drucker-Prager model ``PoroExtendedDruckerPrager`` is used to simulate the elastoplastic behavior of ``rock``. And the single-phase fluid model ``CompressibleSinglePhaseFluid`` is employed to handle the storage and flow of ``water``: 
```

.. literalinclude:: ../../../../../examples/plasticity/WellboreProblem_PoroDruckerPrager.xml
    :language: xml
    :start-after: <!-- SPHINX_WELLBORE_MATERIAL -->
    :end-before: <!-- SPHINX_WELLBORE_MATERIAL_END -->

As for the material parameters, ``defaultInitialFrictionAngle``, ``defaultResidualFrictionAngle`` and ``defaultCohesion`` denote the initial friction angle, the residual friction angle, and cohesion, respectively, as defined by the Mohr-Coulomb failure envelope in the :math:`\Sigma` - :math:`\Tau` plane.

> As the residual friction angle ``defaultResidualFrictionAngle`` is larger than the initial one ``defaultInitialFrictionAngle``, the strain hardening model is automatically chosen, whose hardening rate is given as ``defaultHardening="0.01"``. 

Reviewer: ```suggestion
As the residual friction angle ``defaultResidualFrictionAngle`` is larger than the initial one ``defaultInitialFrictionAngle``, a strain hardening model is automatically chosen, whose hardening rate is given as ``defaultHardening="0.01"``. 
```

If the residual friction angle is set to be less than the initial one, strain weakening will take place. 

> And ``defaultDilationRatio="1.0"`` corresponds to an associated flow rule.

Reviewer: Don't begin sentence with And.

For this coupled problem, the Biot’s coefficient ``BiotCoefficient`` is assumed to be 1.

> If using an incompressible fluid, user can lower the fluid compressibility ``compressibility`` to 0.

Reviewer: ```suggestion
If using an incompressible fluid, the user can lower the fluid compressibility ``compressibility`` to 0.
```

> The constitutive parameters such as the density, the bulk modulus, and the shear modulus are specified in the International System of Units. And stress-dependent porosity model ``rockPorosity`` and constant permeability ``rockPerm`` model are defined in this section.

Reviewer: ```suggestion
The constitutive parameters such as the density, the bulk modulus, and the shear modulus are specified in the International System of Units. A stress-dependent porosity model ``rockPorosity`` and constant permeability ``rockPerm`` model are defined in this section.
```

    :end-before: <!-- SPHINX_WELLBORE_BC_END -->

With ``tractionType="normal"``, traction is applied to the wellbore wall ``rneg`` as a pressure specified from the product of scale ``scale="-10.0e6"`` and the outward face normal. 
A table function ``timeFunction`` is used to define the time-dependent loading. 
The ``coordinates`` and ``values`` form a time-magnitude
pair for the loading time history. In this case, the loading magnitude decreases linearly as the time evolves. 

.. literalinclude:: ../../../../../examples/plasticity/WellboreProblem_PoroDruckerPrager.xml
    :language: xml
    :start-after: <!-- SPHINX_WELLBORE_TABLE -->
    :end-before: <!-- SPHINX_WELLBORE_TABLE_END -->

You may note :

 - All initial value fields must have ``initialCondition`` field set to ``1``;
 - The ``setName`` field points to the previously defined box to apply the fields;
 - ``nodeManager`` and ``faceManager`` in the ``objectPath`` indicate that the boundary conditions are applied to the element nodes and faces, respectively;
 - ``fieldName`` is the name of the field registered in GEOSX;
 - Component ``0``, ``1``, and ``2`` refer to the x, y, and z direction, respectively;

>  - And the non-zero values given by ``Scale`` indicate the magnitude of the loading;

Reviewer: ```suggestion
 - And the non-zero values given by ``scale`` indicate the magnitude of the loading;
```

+------------------+-------------------------+------------------+--------------------+
| :math:`P_wf`     | Fluid Pressure at Well  | [MPa]            | 10.0               |
+------------------+-------------------------+------------------+--------------------+
| :math:`\rho_f`   | Fluid Density           | [kg/m\ :sup:`3`] | 1000.0             |
+------------------+-------------------------+------------------+--------------------+
| :math:`\mu`      | Fluid Viscosity         | [Pa s]           | 0.001              |
+------------------+-------------------------+------------------+--------------------+
| :math:`c_f`      | Fluid Compressibility   | [Pa\ :sup:`-1`]  | 2.09*10\ :sup:`-10`|
+------------------+-------------------------+------------------+--------------------+
| :math:`\kappa`   | Matrix Permeability     | [m\ :sup:`2`]    | 1.0*10\ :sup:`-20` |
+------------------+-------------------------+------------------+--------------------+
| :math:`\phi`     | Porosity                | [-]              | 0.3                |
+------------------+-------------------------+------------------+--------------------+

Inspecting results
---------------------------------

> As defined in the ``Events`` section, we run this simulation for 497640 seconds. In the above examples, we requested silo-format output files. We can therefore import these into VisIt and use python scripts to visualize the outcome. 

Reviewer: Add a note that a non-dimensional time is used in the analytical solution, and the end time here leads to a non-dimensional end time of t* = 1.0 (or whatever).

Author: will add this

.. _problemVerificationFig2:
.. figure:: PoroElastic_X.png
   :align: center
   :width: 1000
   :figclass: align-center

   Fig.4 Comparing GEOSX results with analytical solutions

For the same 3D wellbore problem, the poroplastic case is thereafter tested and compared with the poroelastic one. Fig.5 shows the distribution of :math:`\Sigma_yy` in the near wellbore region for both cases. As observed, a relaxation of the tangential stress along the direction of minimum horizontal stress is detected for the poroplastic case, which can be attributed to the plastic response of the rock.

.. _problemVerificationFig3:
.. figure:: PoroDP_Stress.png
   :align: center
   :width: 1000
   :figclass: align-center

   Fig.5 Simulation result of Syy: PoroElastic vs. PoroPlastic

> By using python scripts, we can extract the simulation results along any direction and provide detailed comparisons between different cases. Here, the pore pressure, radial displacement, radial and tangential effective stresses along the direction of minimum horizontal stress are obtained at different time steps and plotted against the corresponding ones of the poroelastic case. Because of fluid diffusion and coupling effect, Fig.6 shows that these solutions evolve with time for both cases. As mentioned above, plastic zone is developed in the vicinity of the wellbore, due to stress concentration. As for the far field region, these two cases become almost identical, as the rock deformation is governed by poroelasticity.    

Reviewer: ```suggestion
By using python scripts, we can extract the simulation results along any direction and provide detailed comparisons between different cases. Here, the pore pressure, radial displacement, radial and tangential effective stresses along the direction of minimum horizontal stress are obtained at different time steps and plotted against the corresponding ones of the poroelastic case. Because of fluid diffusion and coupling effect, Fig.6 shows that these solutions evolve with time for both cases. As mentioned above, a plastic zone is developed in the vicinity of the wellbore, due to stress concentration. As for the far field region, these two cases become almost identical, with the rock deformation governed by poroelasticity.    
```

.. _problemVerificationFig2:
.. figure:: PoroDP_X.png
   :align: center
   :width: 1000
   :figclass: align-center

>    Fig.6 Comparing PoroPlastic case with PoroElastic case at different time

Reviewer: ```suggestion
   Fig.6 Comparing the PoroPlastic case with the PoroElastic case at different times
```

.. _TutorialPoroPlasticity:

####################################################
Tutorial 12: PoroMechanical Models for Wellbore Problems
####################################################

**Context**

> The main objective of this tutorial is to demonstrate how to use internal mesh generator and poromechanical solvers in GEOSX to tackle wellbore problems in porous media. In this tutorial, a poroplastic model is applied to find the solution of rock deformation within the vicinity of a vertical wellbore, considering elastoplastic deformation, fluid diffusion and poromechanical coupling effect. To do so, a single phase flow solver is fully coupled with a small-strain Lagrangian mechanics solver in GEOSX and the Extended Drucker-Prager model (see :ref:`TwoInvariantPlasticity`) is chosen as the material model for the solid domain. We first solve this problem with the existing poroelastic solver and verify the modeling results with the corresponding analytical solutions. Then, the verified case is employed to test the newly implemented poroplasticity solver in GEOSX, whose results are compared with the ones obtained from the poroelastic case to highlight the impact of plasticity in this specific problem.

Reviewer 2: What is added to this tutorial comparing to the 11st?

Author: This tutorial focuses on the poromechanical solvers (poroelastic and poroplastic) and coupling effect, which is not covered in the 11th one.
